authority/admin/api/webhook.go

```diff
@@ -0,0 +1,200 @@
+// responses.
+type WebhookAdminResponder interface {
+	CreateProvisionerWebhook(w http.ResponseWriter, r *http.Request)
+	UpdateProvisionerWebhook(w http.ResponseWriter, r *http.Request)
+	DeleteProvisionerWebhook(w http.ResponseWriter, r *http.Request)
+}
+
+// webhoookAdminResponder implements WebhookAdminResponder
+type webhookAdminResponder struct{}
+
+// NewWebhookAdminResponder returns a new WebhookAdminResponder
+func NewWebhookAdminResponder() WebhookAdminResponder {
+	return &webhookAdminResponder{}
+}
+
+func validateWebhook(webhook *linkedca.Webhook) error {
+	if webhook == nil {
+		return nil
+	}
+
+	// name
+	if webhook.Name == "" {
+		return admin.NewError(admin.ErrorBadRequestType, "webhook name is required")
+	}
+
+	// url
+	parsedURL, err := url.Parse(webhook.Url)
+	if err != nil {
+		return admin.NewError(admin.ErrorBadRequestType, "webhook url is invalid")
+	}
+	if parsedURL.Host == "" {
+		return admin.NewError(admin.ErrorBadRequestType, "webhook url is invalid")
+	}
+	if parsedURL.Scheme != "https" {
+		return admin.NewError(admin.ErrorBadRequestType, "webhook url must use https")
+	}
+	if parsedURL.User != nil {
+		return admin.NewError(admin.ErrorBadRequestType, "webhook url may not contain username or password")
+	}
+
+	// kind
+	if webhook.Kind != linkedca.Webhook_ENRICHING {
+		return admin.NewError(admin.ErrorBadRequestType, "webhook kind is invalid")
+	}
+
+	return nil
+}
+
+func (war *webhookAdminResponder) CreateProvisionerWebhook(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	prov := linkedca.MustProvisionerFromContext(ctx)
+
+	var newWebhook = new(linkedca.Webhook)
+	if err := read.ProtoJSON(r.Body, newWebhook); err != nil {
+		render.Error(w, err)
+		return
+	}
+
+	if err := validateWebhook(newWebhook); err != nil {
+		render.Error(w, err)
+		return
+	}
+
+	id, err := randutil.UUIDv4()
+	if err != nil {
+		render.Error(w, admin.WrapErrorISE(err, "error generating webhook id"))
+		return
+	}
+	newWebhook.Id = id
+
+	// verify the name is unique
+	for _, wh := range prov.Webhooks {
+		if wh.Name == newWebhook.Name {
+			err := admin.NewError(admin.ErrorConflictType, "provisioner %s already has a webhook with the name %s", prov.Name, newWebhook.Name)
+			render.Error(w, err)
+			return
+		}
+	}
+
+	secret, err := randutil.Bytes(64)
+	if err != nil {
+		render.Error(w, admin.WrapErrorISE(err, "error generating webhook secret"))
+		return
+	}
+	newWebhook.Secret = base64.StdEncoding.EncodeToString(secret)
+
+	prov.Webhooks = append(prov.Webhooks, newWebhook)
+
+	auth := mustAuthority(ctx)
+	if err := auth.UpdateProvisioner(ctx, prov); err != nil {
+		if isBadRequest(err) {
+			render.Error(w, admin.WrapError(admin.ErrorBadRequestType, err, "error creating provisioner webhook"))
+			return
+		}
+
+		render.Error(w, admin.WrapErrorISE(err, "error creating provisioner webhook"))
+		return
+	}
+
+	render.ProtoJSONStatus(w, newWebhook, http.StatusCreated)
```

Comment:
Should we return the secret?

Reply:
How else would they get the secret if we don't return it here?